src/boot/efi/meson.build

```diff
@@ -71,12 +70,106 @@ if efi_conf.get('SBAT_DISTRO', '') != ''
                 section : 'UEFI')
 endif
 
-efi_config_h = configure_file(
+configure_file(
         output : 'efi_config.h',
         configuration : efi_conf)
 
 ############################################################
 
+efi_includes = [fundamental_include, include_directories('.')]
+
+efi_c_args = [
+        '-DSD_BOOT=1',
+        '-ffreestanding',
+        '-fno-strict-aliasing',
+        '-fshort-wchar',
+        '-include', 'efi_config.h',
+]
+
+# Debug information has little value in release builds as no normal human being knows
+# how to attach a debugger to EFI binaries running on real hardware. Anyone who does
+# certainly has the means to do their own dev build.
+if get_option('mode') == 'developer' and get_option('debug')
+        efi_c_args += '-DEFI_DEBUG'
+endif
+
+efi_c_args += cc.get_supported_arguments(
+        '-fwide-exec-charset=UCS2',
+        # gcc docs says this is required for ms_abi to work correctly.
+        '-maccumulate-outgoing-args',
+)
+
+# efi_c_args is explicitly passed to targets so that they can override distro-provided flags
+# that should not be used for EFI binaries.
+efi_c_args += cc.get_supported_arguments(
+        '-fcf-protection=none',
+        '-fno-asynchronous-unwind-tables',
+        '-fno-exceptions',
+        '-fno-sanitize=all',
+        '-fno-stack-clash-protection',
+        '-fno-stack-protector',
+        '-fno-unwind-tables',
+)
+
+efi_c_ld_args = [
+        # We only support bfd. gold is going away, lld has issues with LTO on x86
+        # and mold does not support linker scripts.
+        '-fuse-ld=bfd',
+
+        '-nostdlib',
+        '-static-pie',
+        '-Wl,--entry=efi_main',
+        '-lgcc',
+
+        # These flags should be passed by -static-pie, but seem to be missing sometimes.
+        '-Wl,--no-dynamic-linker',
+        '-z', 'text',
+
+        # EFI has 4KiB pages.
+        '-z', 'common-page-size=4096',
+        '-z', 'max-page-size=4096',
+
+        '-z', 'noexecstack',
+        '-z', 'norelro',
+        '-T' + project_source_root / 'tools/elf2efi.lds',
```

Comment:
I think it'd be more in line with the rest of our config to define `el2efi_lds = files('tools/elf2efi.lds')` in the top-level `meson.build` and reuse it everywhere.